### ISSUE TITLE
Unable to compile with Clang on Windows 10

### ISSUE THREAD

**Issue by theakman2** [jedisct1/libsodium#648]
Hi there,

I'm trying to compile an application against libsodium using Clang 5.0.0 on Windows 10 but I'm running into difficulty compiling the following program...

```c
// test.c
#include <sodium.h>
#include <immintrin.h>
```

...with this command:

`clang -c test.c`

I get the following output:

```
C:\Temp>clang -c test.c
In file included from test.c:2:
In file included from C:\Program Files\LLVM\lib\clang\5.0.0\include\immintrin.h:28:
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:81:40: error: passing '__v2si' (aka 'int') to parameter of
      incompatible type '__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return __builtin_ia32_vec_ext_v2si((__v2si)__m, 0);
                                       ^~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:143:43: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_packsswb((__v4hi)__m1, (__v4hi)__m2);
                                          ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:173:43: error: passing '__v2si' (aka 'int') to parameter of
      incompatible type '__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return (__m64)__builtin_ia32_packssdw((__v2si)__m1, (__v2si)__m2);
                                          ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:203:43: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_packuswb((__v4hi)__m1, (__v4hi)__m2);
                                          ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:230:44: error: passing '__v8qi' (aka 'char') to parameter of
      incompatible type '__attribute__((__vector_size__(8 * sizeof(char)))) char' (vector of 8 'char' values)
    return (__m64)__builtin_ia32_punpckhbw((__v8qi)__m1, (__v8qi)__m2);
                                           ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:253:44: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_punpckhwd((__v4hi)__m1, (__v4hi)__m2);
                                           ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:274:44: error: passing '__v2si' (aka 'int') to parameter of
      incompatible type '__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return (__m64)__builtin_ia32_punpckhdq((__v2si)__m1, (__v2si)__m2);
                                           ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:301:44: error: passing '__v8qi' (aka 'char') to parameter of
      incompatible type '__attribute__((__vector_size__(8 * sizeof(char)))) char' (vector of 8 'char' values)
    return (__m64)__builtin_ia32_punpcklbw((__v8qi)__m1, (__v8qi)__m2);
                                           ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:324:44: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_punpcklwd((__v4hi)__m1, (__v4hi)__m2);
                                           ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:345:44: error: passing '__v2si' (aka 'int') to parameter of
      incompatible type '__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return (__m64)__builtin_ia32_punpckldq((__v2si)__m1, (__v2si)__m2);
                                           ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:366:40: error: passing '__v8qi' (aka 'char') to parameter of
      incompatible type '__attribute__((__vector_size__(8 * sizeof(char)))) char' (vector of 8 'char' values)
    return (__m64)__builtin_ia32_paddb((__v8qi)__m1, (__v8qi)__m2);
                                       ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:387:40: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_paddw((__v4hi)__m1, (__v4hi)__m2);
                                       ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:408:40: error: passing '__v2si' (aka 'int') to parameter of
      incompatible type '__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return (__m64)__builtin_ia32_paddd((__v2si)__m1, (__v2si)__m2);
                                       ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:430:41: error: passing '__v8qi' (aka 'char') to parameter of
      incompatible type '__attribute__((__vector_size__(8 * sizeof(char)))) char' (vector of 8 'char' values)
    return (__m64)__builtin_ia32_paddsb((__v8qi)__m1, (__v8qi)__m2);
                                        ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:453:41: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_paddsw((__v4hi)__m1, (__v4hi)__m2);
                                        ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:475:42: error: passing '__v8qi' (aka 'char') to parameter of
      incompatible type '__attribute__((__vector_size__(8 * sizeof(char)))) char' (vector of 8 'char' values)
    return (__m64)__builtin_ia32_paddusb((__v8qi)__m1, (__v8qi)__m2);
                                         ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:497:42: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_paddusw((__v4hi)__m1, (__v4hi)__m2);
                                         ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:518:40: error: passing '__v8qi' (aka 'char') to parameter of
      incompatible type '__attribute__((__vector_size__(8 * sizeof(char)))) char' (vector of 8 'char' values)
    return (__m64)__builtin_ia32_psubb((__v8qi)__m1, (__v8qi)__m2);
                                       ^~~~~~~~~~~~
C:\Program Files\LLVM\lib\clang\5.0.0\include\mmintrin.h:539:40: error: passing '__v4hi' (aka 'short') to parameter of
      incompatible type '__attribute__((__vector_size__(4 * sizeof(short)))) short' (vector of 4 'short' values)
    return (__m64)__builtin_ia32_psubw((__v4hi)__m1, (__v4hi)__m2);
                                       ^~~~~~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

I believe I've traced the problem to the `export.h` file:

```c
#ifndef sodium_export_H
#define sodium_export_H

#ifndef __GNUC__
# ifdef __attribute__
#  undef __attribute__
# endif
// If I comment out the line below, the errors go away and the file compiles successfully.
//# define __attribute__(a)
#endif
// (snip)
```

The problem seems to be that, on Windows, Clang does not define `__GNUC__`. It's therefore falling into this conditional and `__attribute__` is getting undefined (note that Clang on Windows *does* define `__clang__`, `_WIN32`, `_WIN64` and `_MSC_VER`). Unfortunately `mmintrin.h` (an include from within `immintrin.h`) has the following code at the top:

```c
typedef long long __m64 __attribute__((__vector_size__(8)));

typedef long long __v1di __attribute__((__vector_size__(8)));
typedef int __v2si __attribute__((__vector_size__(8)));
typedef short __v4hi __attribute__((__vector_size__(8)));
typedef char __v8qi __attribute__((__vector_size__(8)));
```

In other words, `immintrin.h` needs the original `__attribute__` to work.

Modifying `export.h` to the following fixes the problem for me. Is this a viable solution?

```c
#ifndef sodium_export_H
#define sodium_export_H

#ifndef __GNUC__
#if !(defined(_MSC_VER) && defined(__clang__))
# ifdef __attribute__
#  undef __attribute__
# endif
# define __attribute__(a)
#endif
#endif
```

**Comment by jedisct1** [comment]
This should be fixed in the `stable` branch.

But libsodium has never been tested on Clang/Windows, so use at your own risk!

**Comment by theakman2** [comment]
Great, thanks!

________________________________
